crates/typst/src/export/svg.rs

```diff
@@ -270,42 +362,149 @@ impl SVGRenderer {
+            );
         } else {
             self.xml.write_attribute("fill", "none");
         }
 
         if let Some(stroke) = &shape.stroke {
-            self.write_stroke(stroke);
+            self.write_stroke(
+                stroke,
+                self.shape_fill_size(state, &stroke.paint, shape),
+                self.shape_paint_transform(state, &stroke.paint, shape),
+            );
         }
 
         let path = convert_geometry_to_path(&shape.geometry);
         self.xml.write_attribute("d", &path);
         self.xml.end_element();
     }
 
+    /// Calculate the transform of the shape's fill or stroke.
+    fn shape_paint_transform(
+        &self,
+        state: State,
+        paint: &Paint,
+        shape: &Shape,
+    ) -> Transform {
+        let mut shape_size = shape.geometry.size();
+        // Edge cases for strokes.
+        if shape_size.x.to_pt() == 0.0 {
+            shape_size.x = Abs::pt(1.0);
+        }
+
+        if shape_size.y.to_pt() == 0.0 {
+            shape_size.y = Abs::pt(1.0);
+        }
+
+        if let Paint::Gradient(gradient) = paint {
+            match gradient.unwrap_relative(false) {
+                Relative::This => {
+                    Transform::scale(shape_size.x.into(), shape_size.y.into())
+                }
+                Relative::Parent => {
+                    Transform::scale(state.size.x.into(), state.size.y.into())
+                        .post_concat(state.transform.invert())
+                }
+            }
+        } else {
+            Transform::identity()
+        }
+    }
+
+    /// Calculate the size of the shape's fill.
+    fn shape_fill_size(&self, state: State, paint: &Paint, shape: &Shape) -> Size {
+        let mut shape_size = shape.geometry.size();
+        // Edge cases for strokes.
+        if shape_size.x.to_pt() == 0.0 {
+            shape_size.x = Abs::pt(1.0);
+        }
+
+        if shape_size.y.to_pt() == 0.0 {
+            shape_size.y = Abs::pt(1.0);
+        }
+
+        if let Paint::Gradient(gradient) = paint {
+            match gradient.unwrap_relative(false) {
+                Relative::This => shape_size,
+                Relative::Parent => state.size,
+            }
+        } else {
+            shape_size
+        }
+    }
+
     /// Write a fill attribute.
-    fn write_fill(&mut self, fill: &Paint) {
-        let Paint::Solid(color) = fill;
-        self.xml.write_attribute("fill", &color.encode());
+    fn write_fill(&mut self, fill: &Paint, size: Size, transform: Transform) {
+        match fill {
+            Paint::Solid(color) => self.xml.write_attribute("fill", &color.encode()),
+            Paint::Gradient(gradient) => {
+                let id = self.push_gradient(gradient, size, transform);
+                self.xml.write_attribute_fmt("fill", format_args!("url(#{id})"));
+            }
+        }
+    }
+
+    /// Pushes a gradient to the SVG file. If the gradient is already present,
```

Comment:
It's not really pushed to `SVG file` right? More to an intermediate structure and it will be written later.